packages/opentelemetry-plugin-xml-http-request/src/xhr.ts

```diff
@@ -0,0 +1,463 @@
+import * as types from '@opentelemetry/types';
+import {
+  addSpanNetworkEvent,
+  getResource,
+  parseUrl,
+  PerformanceTimingNames as PTN,
+} from '@opentelemetry/web';
+import * as shimmer from 'shimmer';
+import { AttributeNames } from './enums/AttributeNames';
+import { EventNames } from './enums/EventNames';
+import { Format } from './enums/Format';
+import {
+  OpenFunction,
+  PropagateTraceHeaderCorsUrls,
+  SendBody,
+  SendFunction,
+  XhrMem,
+} from './types';
+
+/**
+ * XMLHttpRequest config
+ */
+export interface XMLHttpRequestPluginConfig extends types.PluginConfig {
+  // urls which should include trace headers when origin doesn't match
+  propagateTraceHeaderCorsUrls?: PropagateTraceHeaderCorsUrls;
+}
+
+/**
+ * This class represents a XMLHttpRequest plugin for auto instrumentation
+ */
+export class XMLHttpRequestPlugin extends BasePlugin<XMLHttpRequest> {
+  readonly component: string = 'xml-http-request';
+  // @TODO align this with all packages #600
+  readonly version: string = '0.3.0';
+  moduleName = this.component;
+
+  protected _config!: XMLHttpRequestPluginConfig;
+
+  private _tasksCount = 0;
+  private _xhrMem = new WeakMap<XMLHttpRequest, XhrMem>();
+  private _usedResources: { [key: string]: PerformanceResourceTiming[] } = {};
+
+  constructor(config: XMLHttpRequestPluginConfig = {}) {
+    super();
+    this._config = config;
+  }
+
+  /**
+   * Adds custom headers to XMLHttpRequest
+   * @param xhr
+   * @param span
+   * @private
+   */
+  private _addHeaders(xhr: XMLHttpRequest, span: types.Span) {
+    let propagateTraceHeaderUrls =
+      this._config.propagateTraceHeaderCorsUrls || [];
+    if (
+      typeof propagateTraceHeaderUrls === 'string' ||
+      propagateTraceHeaderUrls instanceof RegExp
+    ) {
+      propagateTraceHeaderUrls = [propagateTraceHeaderUrls];
+    }
+
+    const url = (span as tracing.Span).name;
+    const spanUrl = parseUrl(url);
+    let addHeaderOnDifferentOrigin = false;
+
+    if (spanUrl.origin !== window.location.origin) {
+      for (const propagateTraceHeaderUrl of propagateTraceHeaderUrls) {
+        if (urlMatches(url, propagateTraceHeaderUrl)) {
+          addHeaderOnDifferentOrigin = true;
+          break;
+        }
+      }
+      if (!addHeaderOnDifferentOrigin) {
+        this._logger.warn('Cannot set headers on different origin');
+        return;
+      }
+    }
+    const headers: { [key: string]: unknown } = {};
+    this._tracer
+      .getHttpTextFormat()
+      .inject(span.context(), Format.HTTP, headers);
+
+    Object.keys(headers).forEach(key => {
+      xhr.setRequestHeader(key, String(headers[key]));
+    });
+  }
+
+  private _addChildSpan(
+    span: tracing.Span,
+    corsPreFlightRequest: PerformanceResourceTiming
+  ): void {
+    const childSpan = this._tracer.startSpan('CORS Preflight', {
+      startTime: corsPreFlightRequest[PTN.FETCH_START],
+      parent: span,
+    }) as tracing.Span;
```

Comment:
why is this cast needed?

Reply:
because I need tracing.Span not types.Span

Reply:
hmm. that's a problem since the plugins need to be able to be used by sdk that isn't ours as far as I know. If we are looking for things specific to our sdk then we break the api/sdk separation

Reply:
the problem lays somewhere else which I already mentioned with Dave. This is BasicTracer that returns wrong type

Reply:
The `BasicTracer` doesn't come into play here. `this._tracer` is of type `types.Tracer` and `this._tracer.startSpan` returns `types.Span`. The plugin should not care what specific implementation of span is returned from `startSpan` as long as it fulfills the API type.

Can you link the message thread with dave? I can't seem to find it.

Reply:
`https://github.com/open-telemetry/opentelemetry-js/blob/b3332e34b55b67e3aa59c393ab16da84d74fd8a2/packages/opentelemetry-tracing/src/BasicTracer.ts#L72`
`https://github.com/open-telemetry/opentelemetry-js/blob/b3332e34b55b67e3aa59c393ab16da84d74fd8a2/packages/opentelemetry-tracing/src/BasicTracer.ts#L97` which implements
`https://github.com/open-telemetry/opentelemetry-js/blob/b3332e34b55b67e3aa59c393ab16da84d74fd8a2/packages/opentelemetry-tracing/src/Span.ts#L32`


now try to use `span.startTime` as startTime is not part of `types.Span` the returned type from `BasicTracer` form functions `startSpan` or `getCurrentSpan` etc.

Reply:
I started trying to remove the casts on my local copy so I could understand the issue more clearly and i found a couple things:

1. you're accessing `span.name` and `span.attributes` in the plugin. Neither of these properties are on the `Span` API type, so they will not be compatible with SDKs other than ours. I suggest you store what you need in the weak map storage object, so that you don't need to reach back into non-api parts of the span to retrieve it.
2. You're passing `span` into `getAttributes` in `@opentelemetry/web`. Now the plugin depends directly on our sdk, which will cause issues when the plugin is used with sdk implementations other than ours. Those sdks will export tracers that return spans that may not be compatible with `getAttributes` and cause throws.

Reply:
How the plugins can be even used without our sdk ?

Reply:
If they only depend on public API methods then interop should not be a problem

Reply:
Could we just put the URL and other data in the `XhrMem` structure directly? It would duplicate the data being on the span but it would allow the plugin to avoid the cast and just use the span API as designed.

Reply:
This is the solution I had in mind as well

Reply:
ok removed the dependency of tracing